server/queue/queue-api/src/test/java/org/apache/james/queue/api/ManageableMailQueueContract.java

```diff
@@ -0,0 +1,642 @@
+        getManageableMailQueue().enQueue(FakeMail.builder()
+            .name(NAME1)
+            .mimeMessage(createMimeMessage())
+            .sender(MailAddressFixture.OTHER_AT_LOCAL)
+            .recipients(MailAddressFixture.ANY_AT_JAMES, MailAddressFixture.OTHER_AT_JAMES)
+            .lastUpdated(new Date())
+            .build());
+        getManageableMailQueue().enQueue(FakeMail.builder()
+            .name(NAME2)
+            .mimeMessage(createMimeMessage())
+            .sender(MailAddressFixture.OTHER_AT_LOCAL)
+            .recipients(MailAddressFixture.ANY_AT_JAMES, MailAddressFixture.OTHER_AT_JAMES)
+            .lastUpdated(new Date())
+            .build());
+        getManageableMailQueue().enQueue(FakeMail.builder()
+            .name(NAME3)
+            .mimeMessage(createMimeMessage())
+            .sender(MailAddressFixture.OTHER_AT_LOCAL)
+            .recipients(MailAddressFixture.ANY_AT_JAMES, MailAddressFixture.OTHER_AT_JAMES)
+            .lastUpdated(new Date())
+            .build());
+
+        ManageableMailQueue.MailQueueIterator items = getManageableMailQueue().browse();
+        items.next();
+
+        MailQueue.MailQueueItem mailQueueItem = getManageableMailQueue().deQueue();
+
+        assertThat(mailQueueItem.getMail().getName()).isEqualTo(NAME1);
+    }
+
+    @Test
+    default void concurrentEnqueueShouldNotAlterBrowsingWhenDequeueWhileIterating() throws Exception {
+        getManageableMailQueue().enQueue(FakeMail.builder()
+            .name(NAME1)
+            .mimeMessage(createMimeMessage())
+            .sender(MailAddressFixture.OTHER_AT_LOCAL)
+            .recipients(MailAddressFixture.ANY_AT_JAMES, MailAddressFixture.OTHER_AT_JAMES)
+            .lastUpdated(new Date())
+            .build());
+        getManageableMailQueue().enQueue(FakeMail.builder()
+            .name(NAME2)
+            .mimeMessage(createMimeMessage())
+            .sender(MailAddressFixture.OTHER_AT_LOCAL)
+            .recipients(MailAddressFixture.ANY_AT_JAMES, MailAddressFixture.OTHER_AT_JAMES)
+            .lastUpdated(new Date())
+            .build());
+
+        ManageableMailQueue.MailQueueIterator items = getManageableMailQueue().browse();
+
+        ManageableMailQueue.MailQueueItemView firstItem = items.next();
+
+        getManageableMailQueue().enQueue(FakeMail.builder()
+            .name(NAME3)
+            .mimeMessage(createMimeMessage())
+            .sender(MailAddressFixture.OTHER_AT_LOCAL)
+            .recipients(MailAddressFixture.ANY_AT_JAMES, MailAddressFixture.OTHER_AT_JAMES)
+            .lastUpdated(new Date())
+            .build());
+
+        assertThat(firstItem.getMail().getName()).isEqualTo(NAME1);
+        assertThat(items).extracting(ManageableMailQueue.MailQueueItemView::getMail)
+            .extracting(Mail::getName)
+            .containsExactly(NAME2);
+    }
+
+    @Test
+    default void concurrentDequeueShouldNotAlterBrowsingWhileIterating() throws Exception {
```

Comment:
I don't see any iteration here. You mean WhileEnqueueing?

Reply:
You are right, this test is badly named. It should be concurrent**En**queueShouldNotAlterBrowsingWhileIterating